.buildkite/scripts/snyk/report.sh

```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
+
+set -e
+
+MAIN_BRANCH="main"
+cd .buildkite/scripts
+
+# Resolves the branch based on ELASTIC_STACK_VERSION
+# Clones the Logstash repo and builds
+clone_and_build_logstash() {
+  BRANCH="$MAIN_BRANCH"
+  if [ "$ELASTIC_STACK_VERSION" == "$MAIN_BRANCH" ]; then
+    echo "Using $ELASTIC_STACK_VERSION branch."
+  else
+    source snyk/resolve_stack_version.sh
+  fi
+
+  if [ "$ELASTIC_STACK_VERSION" != "$MAIN_BRANCH" ]; then
+    # parse major and minor versions
+    IFS='.'
+    read -a VERSIONS <<< "$ELASTIC_STACK_VERSION"
+
+    BRANCH="${VERSIONS[0]}.${VERSIONS[1]}"
+    echo "Using $BRANCH branch."
+  fi
+
+  git clone --depth 1 --branch "$BRANCH" https://github.com/elastic/logstash.git
+  cd logstash && ./gradlew installDefaultGems && cd ..
+}
+
+# Downloads snyk distribution
+download_snyk() {
+  curl https://static.snyk.io/cli/latest/snyk-linux -o snyk
+  chmod +x ./snyk
+}
+
+# Reports vulnerabilities to the Snyk
+report() {
+  REMOTE_REPO_URL=$MAIN_BRANCH
+  if [ "$ELASTIC_STACK_VERSION" != "$MAIN_BRANCH" ]; then
+    MAJOR_VERSION=$(echo "$ELASTIC_STACK_VERSION"| cut -d'.' -f 1)
+    REMOTE_REPO_URL="$MAJOR_VERSION".latest
+    echo "Using '$REMOTE_REPO_URL' remote repo url."
+  fi
+
+  # TODO: get logstash machine TOKEN from VAULT
+  # ./snyk auth "TOKEN"
+  # ./snyk monitor --all-projects --org=logstash --remote-repo-url="$REMOTE_REPO_URL" --target-reference="$REMOTE_REPO_URL" --detection-depth=10
```

Comment:
@jsvd I am having a difficulties to figure out excluding gem dependencies. `--exclude=DIR,FILE_NAME` option doesn't work because it doesn't accept the structure of `vendor/bundle/jruby/**/gems/*/` ([the Snyk doc ref](https://docs.snyk.io/snyk-cli/commands/monitor#exclude-less-than-name-greater-than-less-than-name-greater-than-...greater-than)). We seem should consider either (not limited to):
- decreasing depth (to 4, as it defaults)
- keep depth=10 and evaluate (skip or accept) in remediation process
Any ideas?

Reply:
maybe we can find a middle ground and determine the depth corresponding to the gems folder under vendor? probably 6 or 7?

Reply:
For my education What is the issue with the bigger depths ? Aren't these GEMs also transitive dependencies of Logstash ?

Reply:
Ruby projects shipped as gems use the .gemspec file as the source of truth for its dependency list and version constraints.
The Gemfile declares dependencies for all kinds of ruby projects, and the Gemfile.lock declares the exact versions to be used, within the constraints declared in the Gemfile.

For a project that is a Ruby Gem, the Gemfile may solely contain:
```ruby
source "http://www.rubygems.org"

gemspec
```
meaning: go find the dependencies on the .gemspec file.

Once the ruby gem is built as a .gem file, the .gemspec file is supposted to be included, but not the Gemfile, and especially not the Gemfile.lock. 
However some gem creators forget to exclude these files (or include them by mistake), so when Snyk looks for a Gemfile+Gemfile.lock pair as an indicator of a project that needs scanning, it finds a few in Logstash:

```
/tmp/logstash-8.7.0
❯ find . -name Gemfile.lock
./Gemfile.lock
./vendor/bundle/jruby/2.6.0/gems/http_parser.rb-0.6.0-java/Gemfile.lock
./vendor/bundle/jruby/2.6.0/gems/pleaserun-0.0.32/Gemfile.lock
./vendor/bundle/jruby/2.6.0/gems/amazing_print-1.4.0/Gemfile.lock
./vendor/bundle/jruby/2.6.0/gems/insist-1.0.0/Gemfile.lock
```

However, these lock files are not used by Logstash at all, and may falsely result in Snyk saying:

"Logstash project has an amazing_print-1.4.0 project, whose Gemfile.lock lists a nokogiri version with a CVE":

```
❯ grep nokogiri  ./vendor/bundle/jruby/2.6.0/gems/amazing_print-1.4.0/Gemfile.lock        
    nokogiri (1.12.5)
  nokogiri (~> 1.10)
```

So ideally we'd like to exclude any Gemfile/Gemfile.lock file from LOGSTASH_HOME/vendor/bundle/jruby/**/gems/*.
The alternative is to tell snyk to not look so deeply into the folder structure, as we know anything found there will be a false positive.

Reply:
Thank you @jsvd that makes total sense, thanks for taking the time to write this down 🙏

Reply:
As we have observed, depth 6 is the target where we can avoid unnecessary (out of Logstash scope) dependencies.